bigtable/google/cloud/bigtable/row_data.py

```diff
@@ -286,6 +300,29 @@ def cancel(self):
         """Cancels the iterator, closing the stream."""
         self._response_iterator.cancel()
 
+    def create_retry_request(self):
+        row_range = self.request_pb.rows.row_ranges.pop()
+        range_kwargs = {}
+        range_kwargs['start_key_open'] = self.last_scanned_row_key
+        range_kwargs['end_key_open'] = row_range.end_key_open
+        self.request_pb.rows.row_ranges.add(**range_kwargs)
+
+    def retry_read_rows(self):
+        if (self.last_scanned_row_key and self.request_pb is not None):
```

Comment:
Can you help me understand the logic here?

Reply:
Please see the comments added on how last_scanned_row_key is used to restart the scan.  Also, the helper methods were made private and docstrings added.

This latest revision has been running for several days on our endurance machine and we have not seen any of the read_rows failures previously seen, so the testing looks good. If you are OK with the changes, can we merge these?

Reply:
> Please see the comments added on how last_scanned_row_key is used to restart the scan. Also, the helper methods were made private and docstrings added.

You added two new args to the constructor and didn't document them. These argument appear to significantly change the behavior of the class. Can you update the docstring to describe both the arguments and the behavior when those arguments are specified?

> This latest revision has been running for several days on our endurance machine and we have not seen any of the read_rows failures previously seen, so the testing looks good. If you are OK with the changes, can we merge these?

That's great, but I want to understand these changes before we merge.

Reply:
Added the additional parameter documentation and usage, sorry for omission. Note that these were added as optional parameters to maintain backward compatibility.